File: GitToolBox/src/main/java/zielu/gittoolbox/metrics/ProjectMetricsImpl.java
package zielu.gittoolbox.metrics;

import com.codahale.metrics.Counter;
import com.codahale.metrics.Gauge;
import com.codahale.metrics.MetricSet;
import com.codahale.metrics.Timer;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectManager;
import com.intellij.openapi.project.ProjectManagerListener;
import com.intellij.openapi.util.Disposer;
import org.jetbrains.annotations.NotNull;
import zielu.gittoolbox.util.DisposeSafeRunnable;

class ProjectMetricsImpl implements ProjectMetrics {
  private final MetricManager metrics = new MetricManager();

  ProjectMetricsImpl(@NotNull Project project) {
    project.getMessageBus().connect(project).subscribe(ProjectManager.TOPIC, new ProjectManagerListener() {
      @Override
      public void projectOpened(@NotNull Project project) {
        ApplicationManager.getApplication()
            .executeOnPooledThread(new DisposeSafeRunnable(project, () -> startReporter(project)));
      }
    });
  }

  private void startReporter(@NotNull Project project) {
    MetricsReporter reporter = Jmx.reporter(project, metrics.getRegistry());
    Disposer.register(project, reporter);
  }

  @Override
  public Timer timer(@NotNull String simpleName) {
    return metrics.timer(simpleName);
  }

  @Override
  public Counter counter(@NotNull String simpleName) {
    return metrics.counter(simpleName);
  }

  @Override
  public <T> Gauge gauge(@NotNull String simpleName, Gauge<T> gauge) {
    return metrics.gauge(simpleName, gauge);
  }

  @Override
  public void addAll(MetricSet metricSet) {
    metrics.addAll(metricSet);
  }
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/metrics/ProjectMetricsImpl.kt
package zielu.gittoolbox.metrics

import com.intellij.openapi.Disposable
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Disposer
import zielu.gittoolbox.GitToolBoxApp
import zielu.gittoolbox.metrics.Jmx.startReporting
import zielu.intellij.concurrent.DisposeSafeRunnable
import zielu.intellij.concurrent.ZDisposableRunnableWrapper
import zielu.intellij.metrics.GtCounter
import zielu.intellij.metrics.GtGauge
import zielu.intellij.metrics.GtTimer
import zielu.intellij.metrics.codehale.CodehaleMetricsManager

internal class ProjectMetricsImpl(
  private val project: Project
) : ProjectMetrics, Disposable {
  private val metrics = CodehaleMetricsManager()

  override fun startReporting() {
    val operation = ZDisposableRunnableWrapper { startReporter(project) }
    Disposer.register(this, operation)
    GitToolBoxApp.getInstance().ifPresent {
      it.runInBackground(DisposeSafeRunnable(operation))
    }
  }

  private fun startReporter(project: Project) {
    Disposer.register(this, startReporting(project, metrics.getRegistry()))
  }

  override fun timer(simpleName: String): GtTimer = metrics.timer(simpleName)

  override fun counter(simpleName: String): GtCounter = metrics.counter(simpleName)

  override fun <T : Any?> gauge(simpleName: String, value: () -> T): GtGauge {
    return metrics.gauge(simpleName, value)
  }

  override fun dispose() {
    // do nothing
  }
}
